appengine/genetics/js/genetics.js

```diff
@@ -237,14 +513,507 @@ Genetics.init = function() {
+              return true;
+            }
+          } else {
+            // If the player had an unsuccessful mating attempt, then the level
+            // is failed.
+            new Genetics.Cage.Event('END_GAME', false).addToQueue();
+            return true;
+          }
+        } else if (event['TYPE'] != 'START_GAME' &&
+            event['TYPE'] != 'NEXT_ROUND') {
+          // If an unexpected event occurred, then the level is failed.
+          new Genetics.Cage.Event('END_GAME', false).addToQueue();
+          return true;
+        }
+      }
+      return false;
+    case 7:
+      // Player was asked to respond to mate request such that they return true
+      // if a mate is valid, false otherwise. The player succeeds if they
+      // correctly respond 5 times.
+      // fallthrough
+    case 8:
+      // Player was asked to respond to mate request such that they return true
+      // if a mate is valid and bigger than itself, false otherwise. The player
+      // succeeds if they correctly respond 5 times.
+      var successfulMates = 0;
+      for (var i = 0, event; event = Genetics.Cage.Events[i]; i++) {
+        if (event['TYPE'] == 'MATE') {
+          if (event['RESULT'] == 'SUCCESS') {
+            if (BlocklyGames.LEVEL == 8 &&
+                Genetics.Cage.miceMap_[event['ID']].size <=
+                Genetics.Mouse.SIZE) {
+              // If it is level 8 and the player tried to mate with someone
+              // that wasn't bigger than itself, they fail.
+              new Genetics.Cage.Event('END_GAME', false).addToQueue();
+              return true;
+            }
+            successfulMates++;
+            if (successfulMates == 5) {
+              new Genetics.Cage.Event('END_GAME', true).addToQueue();
+              return true;
+            }
+          } else if (event['RESULT'] == 'REJECTION') {
+            var self = Genetics.Cage.miceMap_[0];
+            var mate = Genetics.Cage.miceMap_[event['ID']];
+            if (self.sex != mate.sex && (BlocklyGames.LEVEL != 8 ||
+                self.size < mate.sex)) {
+              // If the player rejected a correct mate request, then they fail.
+              new Genetics.Cage.Event('END_GAME', false).addToQueue();
+              return true;
+            }
+          } else {
+            // If the mate event is anything other than a success or rejection,
+            // then the level is failed.
+            new Genetics.Cage.Event('END_GAME', false).addToQueue();
+            return true;
+          }
+        } else if (event['TYPE'] != 'START_GAME' &&
+            event['TYPE'] != 'NEXT_ROUND' && event['TYPE'] != 'RETIRE') {
+          // If an unexpected event occurred, then the level is failed.
+          new Genetics.Cage.Event('END_GAME', false).addToQueue();
+          return true;
+        }
+      }
+      return false;
+    default:
+      // Continue on to handle level 9 or 10.
```

Comment:
It's really a bit unexpected for the logic for levels 9 and 10 to be outside the switch/case. I wonder if it could be here, but maybe pulled out into a function?